contracts/extensions/StreamingPayments.sol

```diff
@@ -0,0 +1,405 @@
+    uint256 _amount
+  )
+    public
+    validateFundingPermission(_fundingPermissionDomainId, _fundingChildSkillIndex, streamingPayments[_id].domainId)
+  {
+    claim(_permissionDomainId, _childSkillIndex, _fromChildSkillIndex, _toChildSkillIndex, _id, toArray(_token));
+
+    PaymentToken storage paymentToken = paymentTokens[_id][_token];
+    require(paymentToken.lastClaimed >= block.timestamp, "streaming-payments-insufficient-funds");
+    paymentToken.amount = _amount;
+
+    emit PaymentTokenUpdated(msgSender(), _id, _token, _amount);
+  }
+
+  /// @notice Update the startTime, only if the current startTime is in the future
+  /// @param _adminPermissionDomainId The domain in which the caller holds the admin permission
+  /// @param _adminChildSkillIndex The index linking the adminPermissionDomainId to the domainId
+  /// @param _id The id of the streaming payment
+  /// @param _startTime The new startTime to set
+  function setStartTime(
+    uint256 _adminPermissionDomainId,
+    uint256 _adminChildSkillIndex,
+    uint256 _id,
+    uint256 _startTime
+  )
+    public
+    validateAdministrationPermission(_adminPermissionDomainId, _adminChildSkillIndex, streamingPayments[_id].domainId)
+
+  {
+    StreamingPayment storage streamingPayment = streamingPayments[_id];
+    require(block.timestamp <= streamingPayment.startTime, "streaming-payments-already-started");
+    streamingPayment.startTime = _startTime;
+  }
+
+  /// @notice Update the endTime, only if the new endTime is in the future
+  /// @param _adminPermissionDomainId The domain in which the caller holds the admin permission
+  /// @param _adminChildSkillIndex The index linking the adminPermissionDomainId to the domainId
+  /// @param _id The id of the streaming payment
+  /// @param _endTime The new endTime to set
+  function setEndTime(
+    uint256 _adminPermissionDomainId,
+    uint256 _adminChildSkillIndex,
+    uint256 _id,
+    uint256 _endTime
+  )
+    public
+    validateAdministrationPermission(_adminPermissionDomainId, _adminChildSkillIndex, streamingPayments[_id].domainId)
+
+  {
+    StreamingPayment storage streamingPayment = streamingPayments[_id];
+    require(block.timestamp <= _endTime, "streaming-payments-invalid-end-time");
+    streamingPayment.endTime = _endTime;
+  }
+
+  /// @notice Cancel the streaming payment, specifically by setting endTime to block.timestamp
+  /// @param _adminPermissionDomainId The domain in which the caller holds the admin permission
+  /// @param _adminChildSkillIndex The index linking the adminPermissionDomainId to the domainId
+  /// @param _id The id of the streaming payment
+  function cancel(
+    uint256 _adminPermissionDomainId,
+    uint256 _adminChildSkillIndex,
+    uint256 _id
+  )
+    public
+    validateAdministrationPermission(_adminPermissionDomainId, _adminChildSkillIndex, streamingPayments[_id].domainId)
+  {
+    setEndTime(_adminPermissionDomainId, _adminChildSkillIndex, _id, block.timestamp);
```

Comment:
I believe the current restriction on `setEndTime` would prevent that behavior -- you couldn't cancel an already cancelled payment, because the new block time would be later than the existing end time.

Reply:
You mean 

```
    require(block.timestamp <= _endTime, "streaming-payments-invalid-end-time");
```

this restriction? Or something else I'm blind to? Because `_endTime` there is the function parameter, which is `block.timestamp` for this call, is it not?

Reply:
Yup, you're right. I was thinking of something different, which was not allowing an endTime to be set if the existing endTime has elapsed (you can't change something which has ended), which would be a new check.